packages/endo/src/infer-exports.js

```diff
@@ -0,0 +1,88 @@
+import { join, relativize } from "./node-module-specifier.js";
+
+const { entries, fromEntries } = Object;
+
+// q, as in quote, for quoting strings in error messages;
+const q = JSON.stringify;
+
+function* interpretBrowserExports(name, exports) {
+  if (typeof exports === "string") {
+    yield [name, relativize(exports)];
+    return;
+  }
+  if (Object(exports) !== exports) {
+    throw new Error(
+      `Cannot interpret package.json browser property for package ${name}, must be string or object, got ${exports}`
+    );
+  }
+  for (const [key, value] of entries(exports)) {
+    yield [join(name, key), relativize(value)];
+  }
+}
+
+function* interpretExports(name, exports, tags) {
+  if (typeof exports === "string") {
+    yield [name, relativize(exports)];
+    return;
+  }
+  if (Object(exports) !== exports) {
+    throw new Error(
+      `Cannot interpret package.json exports property for package ${name}, must be string or object, got ${exports}`
+    );
+  }
+  for (const [key, value] of entries(exports)) {
+    if (key.startsWith("./") || key === ".") {
+      yield* interpretExports(join(name, key), value, tags);
+    } else if (tags.has(key)) {
+      yield* interpretExports(name, value, tags);
+    }
+  }
+}
+
+export function* inferExportsEntries(
```

Comment:
Am I correct in deducing the docstring for this function?:

Given an unpacked `package.json`, generate a series of `[name, target]` pairs to represent what this package exports. `name` is what the caller/importer asked for (e.g. the `ses` in `import { stuff } from 'ses'`, or the `ses/deeper` in `import { stuff } from 'ses/deeper'`). `target` is the path relative to the imported package's root: frequently `./index.js` or `./src/index.js` or (for a deep import) `./src/deeper.js`. There may be multiple pairs for a single `name`, but they will be sorted in ascending priority order, and the caller should use the last one that exists on disk.

Reply:
Oh, the caller is using this array as the argument of `Object.fromEntries`, which presumeably lets the later entries overwrite the earlier ones. So the caller always uses the last entry for any given name.

Nothing here says anything about whether those files exist on disk or not. We assume the `package.json` has an accurate `exports` property which only refers to real files.

Reply:
Adding, with slight amendments:

```js
// Given an unpacked `package.json`, generate a series of `[name, target]`
// pairs to represent what this package exports. `name` is what the
// caller/importer asked for (for example, the `ses` in `import { stuff } from
// 'ses'`, or the `ses/deeper` in `import { stuff } from 'ses/deeper'`).
// `target` is the path relative to the imported package's root: frequently
// `./index.js` or `./src/index.js` or (for a deep import) `./src/deeper.js`.
// There may be multiple pairs for a single `name`, but they will be yielded in
// ascending priority order, and the caller should use the last one that exists.
```

Yielded instead of sorted in order of ascending priority. `fromEntries` respects this semantics, and I *do* feel like a snob for the singular “this” in “this semantics”.

Removed the “on disk” qualification. If the winning entry corresponds to a non-existent file, that is an error in the content of the `package.json`. We will likely need to amend this routine so that it can infer an exports map from the material available on disk if there is not an explicit `exports` field.